snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,387 @@
+    """
+    A ``Consumer`` class that implements an API that is similar to the
+    ``Message``-based API, except that it returns a ``TaskSet`` instance
+    rather than a Kafka ``Message``.
+
+    This consumer uses the progression of time in a Kafka partition as a
+    "virtual clock" rather than depending on wall clock time. This allows
+    replaying history without gaps in task execution if the consumer
+    is not running when a task would have been scheduled to execute for
+    increased reliability (no schedule gaps) and easier testing on historical
+    data.
+
+    This consumer differs from a typical consumer in the way that messages
+    are processed: the tasks that are contained within a task set are
+    identified by whether or not they fall within a time interval formed by
+    the timestamps of the last two messages received within a partition.
+
+    The first message received will not cause a `TaskSet` to be returned by
+    the consumer, since only one endpoint of the interval can be established.
+
+    The second message received by the consumer will cause a `TaskSet` to be
+    created with all tasks scheduled between the time interval represented by
+    the timestamps of the previous two messages consumed within that
+    partition. (This interval's lower bound is exclusive, and the upper bound
+    is inclusive.) All subsequent messages recieved by the consumer will
+    similarly cause a ``TaskSet`` to be created with all tasks scheduled
+    between the timestamps of the previous two messages consumed within that
+    partition as well.
+
+    This consumer requires that the ``message.timestamp.type`` topic
+    configuration is set to ``LogAppendTime`` for the topics that it is
+    subscribed to, so that that the message time is set by the primary broker
+    for the topic -- not the producer of the message -- ensuring that each
+    partition timestamp moves monotonically.
+
+    There is also a slight difference between the vanilla ``Consumer``
+    behavior and this implementation during rebalancing: whenever a partition
+    is assigned to this consumer, offsets are *always* automatically reset to
+    the committed offset for that partition (or if no offsets have been
+    committed for that partition, the offset is reset in accordance with the
+    ``auto.offset.reset`` configuration value.) This causes partitions that
+    are maintained across a rebalance to have the same offset management
+    behavior as a partition that is moved from one consumer to another. To
+    prevent uncommitted messages from being consumed multiple times,
+    ``commit`` should be called in the partition revocation callback.
+
+    This consumer can either be used with automatic timed offset commit (the
+    librdkafka default, set using the ``enable.auto.commit`` configuration
+    value) or manual calls to ``commit``.
+    """
+
+    # Since this consumer deals with time intervals between messages rather
```

Comment:
I wanted to split this in a way where the docstring is the things that you should know if you are using the consumer, and the comment is things you should know if you're editing the consumer itself. There is a little bit of redundancy but I think that is okay (and I think with the subtle complexities here it's better to err on the side of being too thorough rather than leaving out detail.) With that in mind, are there still changes that you think make sense here? I'm open to make some, I think I'm just too entrenched in the details to see what would make sense to change here.